web/src/main/java/com/google/accompanist/web/WebView.kt

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright 2021 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.accompanist.web
+
+import android.graphics.Bitmap
+import android.net.Uri
+import android.webkit.WebResourceRequest
+import android.webkit.WebView
+import android.webkit.WebViewClient
+import androidx.activity.compose.BackHandler
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.viewinterop.AndroidView
+
+/**
+ * A wrapper around the Android View WebView to provide a basic WebView composable.
+ *
+ * If you require more customisation you are most likely better rolling your own and using this
+ * wrapper as an example.
+ *
+ * @param state The webview state holder where the Uri to load is defined.
+ * @param captureBackPresses Set to true to have this Composable capture back presses and navigate
+ * the WebView back.
+ * @param onCreated Called when the WebView is first created, this can be used to set additional
+ * settings on the WebView.
+ *
+ * @sample com.google.accompanist.sample.webview.BasicWebViewSample
+ */
+@Composable
+fun WebView(
+    state: WebViewState,
+    onContentChanged: (WebContent) -> Unit,
+    modifier: Modifier = Modifier,
+    captureBackPresses: Boolean = true,
+    onCreated: (WebView) -> Unit = {}
+) {
+    var view by remember { mutableStateOf<WebView?>(null) }
+    var webViewLoaded by remember(view) { mutableStateOf(false) }
+
+    BackHandler(captureBackPresses && state.canGoBack) {
+        view?.goBack()
+    }
+
+    AndroidView(
+        factory = {
+            WebView(it).apply {
+                onCreated(this)
+
+                webViewClient = object : WebViewClient() {
+                    override fun onPageStarted(view: WebView?, url: String?, favicon: Bitmap?) {
+                        super.onPageStarted(view, url, favicon)
+                        state.isLoading = true
+
+                        // When a WebView is first created it sends a page started event
+                        // with about:blank. We don't want to forward this on.
+                        // It's important we still support a link to about:blank though.
+                        if (webViewLoaded || url != "about:blank") {
```

Comment:
I'm not familiar with the WebViewClient callback but.. can `onPageStarted` be called when `webViewLoaded == true`? In that case, why does it notified that the content changed, but you don't change the content?

Reply:
This was to get around webview updating the url itself sometimes like in the case of redirects and navigating back. I have found a better way to do it though that removes the need for the state update here.